Sources/NIOCore/AsyncSequences/NIOBackPressuredAsyncSequence.swift

```diff
@@ -0,0 +1,1068 @@
+            /// Indicates that the continuation should be resumed and
+            /// ``NIOBackPressuredAsyncSequence/Source/YieldResult/produceMore`` should be returned.
+            case resumeContinuationAndReturnProduceMore(
+                continuation: CheckedContinuation<Element?, Never>,
+                element: Element
+            )
+            /// Indicates that the continuation should be resumed and
+            /// ``NIOBackPressuredAsyncSequence/Source/YieldResult/stopProducing`` should be returned.
+            case resumeContinuationAndReturnStopProducing(
+                continuation: CheckedContinuation<Element?, Never>,
+                element: Element
+            )
+            /// Indicates that the yielded elements have been dropped.
+            case returnDropped
+
+            @usableFromInline
+            init(shouldProduceMore: Bool, continuationAndElement: (CheckedContinuation<Element?, Never>, Element)? = nil) {
+                switch (shouldProduceMore, continuationAndElement) {
+                case (true, .none):
+                    self = .returnProduceMore
+
+                case (false, .none):
+                    self = .returnStopProducing
+
+                case (true, .some((let continuation, let element))):
+                    self = .resumeContinuationAndReturnProduceMore(
+                        continuation: continuation,
+                        element: element
+                    )
+
+                case (false, .some((let continuation, let element))):
+                    self = .resumeContinuationAndReturnStopProducing(
+                        continuation: continuation,
+                        element: element
+                    )
+                }
+            }
+        }
+
+        @inlinable
+        mutating func yield<S: Sequence>(_ sequence: S) -> YieldAction where S.Element == Element {
+            switch self.state {
+            case .initial(var backPressureStrategy, let iteratorInitialized):
+                let buffer = CircularBuffer<Element>(sequence)
+                let shouldProduceMore = backPressureStrategy.didYield(bufferDepth: buffer.count)
+
+                self.state = .streaming(
+                    backPressureStrategy: backPressureStrategy,
+                    buffer: buffer,
+                    continuation: nil,
+                    hasOutstandingDemand: shouldProduceMore,
+                    iteratorInitialized: iteratorInitialized
+                )
+
+                return .init(shouldProduceMore: shouldProduceMore)
+
+            case .streaming(var backPressureStrategy, var buffer, .some(let continuation), _, let iteratorInitialized):
+                // The buffer should always be empty if we hold a continuation
+                precondition(buffer.isEmpty)
+
+                self.state = .modifying
+
+                buffer.append(contentsOf: sequence)
+                let element = buffer.popFirst()
```

Comment:
We could eliminate `buffer.popFirst()` by first getting the first element of the sequence and appending the remaining elements to the buffer. Not straight forward because `Sequence` doesn't guarantee it can be iterated multiple times but possible. I have done something similar for an `AsyncSequence` already:
https://github.com/apple/swift-nio/blob/ece5057615d1bee848341eceafdf04ca54d60177/Sources/NIOCore/AsyncAwaitSupport.swift#L299-L334

This will be especially nice for the case where `sequence.count == 1` as it will eliminate the roundtrip through the circular buffer. Not sure how much performance this will actually bring in practice though, so just a nice to have improvement.

Reply:
While I get what you want to optimise for here this would actually end up in something like this:

```swift
var iterator = sequence.makeIterator()

guard let firstElement = iterator.next() else {
    return .init(shouldProduceMore: hasOutstandingDemand)
}

while let nextElement = iterator.next() {
    buffer.append(nextElement)
}
```

This would actually result in worse code since it would repeatedly call `append`. Every call to append might grow the buffer and it has to adjust the index. Furthermore, for trivial types an implementation of `append(contentsOf:)` could do just a slab copy of the sequence.

After talking with @Lukasa about this a bit, I am going forward with early exiting if the buffer is empty. Furthermore, I am going to adopt the `Deque` from swift-collections since this provides a specialized `append(contentsOf:)` implementation.